### ISSUE TITLE
[Javascript] Adaptive card throws error in parsing json of $data

### ISSUE THREAD

**Issue by atishayv** [microsoft/AdaptiveCards#4781]
# Platform

* JavaScript

# Author or host

Host Name : Microsoft Search

# Version of SDK

"adaptivecards": "2.0.0",

# Details

I am trying to render an adaptive card text block with following json :
```
{
    "type": "TextBlock",
    "text": "${json(message).releaseName}"
}
```

having $data as : 
```
{
    "message": "{\"type\":\"Deployment\",\"buildId\":\"9542982\",\"releaseId\":\"129\",\"buildNumber\":\"20180504.3\",\"releaseName\":\"Release-104\",\"repoProvider\":\"GitHub\"}"
}
```
using the following code :
```
  // Create a Template instance from the template payload
  const template = new ACData.Template(payload);
  // Create a data binding context, and set its $root property to the
  // data object to bind the template to
  const context: ACData.IEvaluationContext = {
    $root: sourceData,
  };
  // "Expand" the template - this generates the final Adaptive Card,
  // ready to render
  const cardPayload = template.expand(context);

   adaptiveCard.parse(cardPayload);
  const adaptiveCardHtml = adaptiveCard.render();

```
but calling render is **throwing error while parsing json saying Unexpected value "\\".**

From designer it works because I think its being handled by monaco editor and designer which parses the $data and remove the backlash from it before giving the data json to adaptive card library.

Can you please look into the issue?

**Comment by atishayv** [comment]
CC: @dclaux

**Comment by shalinijoshi19** [comment]
Hey @dclaux can you pls help take a look at this one for templating? thanks!

**Comment by dclaux** [comment]
@atishayv like you I can't repro in the designer, but the designer isn't doing anything specific - it does basically the exact same thing you do. Can you post a screenshot of the call stack and the error in the browser?

**Comment by atishayv** [comment]
@dclaux Sorry for the late response on this. Not sure how I missed it.
The issue is, we type json in a sample data editor of layout designer and we give the same json to the adaptive card library as well. So, lets consider the json is something like : 
```
{
    "message": "{\"type\":\"Deployment\",\"buildId\":\"9542982\",\"releaseId\":\"129\",\"buildNumber\":\"20180504.3\",\"releaseName\":\"Release-104\",\"repoProvider\":\"GitHub\"}"
}
```
And we give the same value to $data when calling the adaptive card library.

In designer, we do something like this to get the $data value from sample data editor : 
```
t.prototype.updateSampleData = function() {
            try {
                this._sampleData = JSON.parse(this.getCurrentSampleDataEditorPayload()),
                this.scheduleUpdateCardFromJson()
            } catch (e) {}
        }

         t.prototype.getCurrentSampleDataEditorPayload = function() {
            return this._isMonacoEditorLoaded && this._sampleDataEditor ? this._sampleDataEditor.getValue() : ""
        }

```
As you can check, the designer is again parsing the data after getting the value from monaco editor. Its because, monaco editor gives string output even though its a valid json.
So the output from ```this._sampleDataEditor.getValue()``` is :
```
"{
    "message": "{\"type\":\"Deployment\",\"buildId\":\"9542982\",\"releaseId\":\"129\",\"buildNumber\":\"20180504.3\",\"releaseName\":\"Release-104\",\"repoProvider\":\"GitHub\"}",
    
}"
```

and when the designer do a JSON.parse on it, it removes all escape characters. So the final output of ```JSON.parse(this.getCurrentSampleDataEditorPayload())``` is : 
![image](https://user-images.githubusercontent.com/13940020/94798081-b8af3a00-03fe-11eb-85ab-e1cd571f77eb.png)

If you see the message value, the backlash escape character is removed and this is what is given to adaptive card library.

However if we give this json to adaptive card library with escape string which is also a valid json : 

```
{
    "message": "{\"type\":\"Deployment\",\"buildId\":\"9542982\",\"releaseId\":\"129\",\"buildNumber\":\"20180504.3\",\"releaseName\":\"Release-104\",\"repoProvider\":\"GitHub\"}"
}
```
I get following error :
![image](https://user-images.githubusercontent.com/13940020/94784982-98767f80-03ec-11eb-9dbc-dec6e1a1c264.png)

To fix it, I had to do some pre processing over $data to remove the escape characters. Something like this : 
```
$data = JSON.parse(JSON.stringify($data).replace(/(?:\\(.))/g, "$1"));
```

Basically what I am trying to say is passing $data value(which is a valid json) with escape characters to adaptive card library throws error.

Let me know if it clarifies things

**Comment by dclaux** [comment]
Please bear with me as I make sure I understand what you're saying. Essentially, you want your `$data` to be something along those lines:
```
{
    "message": "{\"property\": \"value\"}"
}
```
So essentially, the `message` property represents JSON expressed as a string. Correct?

If so, then again I can't see what's not working as expected. Here is a super simple example I put together in the designer:
![image](https://user-images.githubusercontent.com/1334689/94834563-5dad2100-03c5-11eb-9bbd-f7b386f0f78d.png)

As you can see, the value of the "message" property is correctly displayed in the TextBLock.

For this to happen, the designer sets `_sampleData` in `updateSampleData` as you outlined:
![image](https://user-images.githubusercontent.com/1334689/94835267-56d2de00-03c6-11eb-991c-7fd3fbeae9a3.png)

As you can see, `_sampleData` is correctly parsed as a JSON object with one message property of type string. You have observed exactly the same thing yourself with your own example:
![image](https://user-images.githubusercontent.com/13940020/94798081-b8af3a00-03fe-11eb-85ab-e1cd571f77eb.png)

But the escape (backslash) characters have not been "removed"; it's just VS Code pretty printing (and not showing them - they're just escape characters after all, and their role is to make it possible to correctly interpret things.) You can easily confirm that in the debugger:
![image](https://user-images.githubusercontent.com/1334689/94836489-e4fb9400-03c7-11eb-9abc-785966cb3c41.png)

If I inject your data sample into my minimal template, I get to the same point:
![image](https://user-images.githubusercontent.com/1334689/94836992-85ea4f00-03c8-11eb-8b69-36e3b5aff8ca.png)

And the code the designer uses to bind the template and the data is (most likely) the exact same as yours.

At this point there are two things that need to be clarified to be able to diagnose:
- I don't know what your template looks like
- What version of the templating engine you're using. I'm going to guess you are using the pre-release library while I am using the released one (the designer is too) that is now relying on a more robust expression parser (Adaptive Expressions). It is possible that the pre-release library had a bug that I simply can't repro anymore

**Comment by atishayv** [comment]
Thanks @dclaux for the detailed info. I am not able to reproduce this now. I am not sure which version(probably rc) of library I was using which was causing it.

I am closing this for now. Will follow up on this thread if I am able to reproduce this.
